src/query/stream/streams.cpp

```diff
@@ -210,36 +214,48 @@ Streams::StreamsMap::iterator Streams::CreateConsumer(StreamsMap &map, const std
 
         const static std::map<std::string, storage::PropertyValue> empty_parameters{};
 
-        for (auto &row : result.rows) {
-          spdlog::trace("Processing row in stream '{}'", stream_name);
-          auto [query_value, params_value] =
-              ExtractTransformationResult(std::move(row.values), transformation_name, stream_name);
-          storage::PropertyValue params_prop{params_value};
-
-          std::string query{query_value.ValueString()};
-          spdlog::trace("Executing query '{}' in stream '{}'", query, stream_name);
-          auto prepare_result =
-              interpreter->Prepare(query, params_prop.IsNull() ? empty_parameters : params_prop.ValueMap(), nullptr);
-          if (!interpreter_context->auth_checker->IsUserAuthorized(owner, prepare_result.privileges)) {
-            throw StreamsException{
-                "Couldn't execute query '{}' for stream '{}' becuase the owner is not authorized to execute the "
-                "query!",
-                query, stream_name};
+        for (size_t i = 0; i < total_retries; ++i) {
+          try {
+            for (auto &row : result.rows) {
+              spdlog::trace("Processing row in stream '{}'", stream_name);
+              auto [query_value, params_value] =
+                  ExtractTransformationResult(std::move(row.values), transformation_name, stream_name);
+              storage::PropertyValue params_prop{params_value};
+
+              std::string query{query_value.ValueString()};
+              spdlog::trace("Executing query '{}' in stream '{}'", query, stream_name);
+              auto prepare_result = interpreter->Prepare(
+                  query, params_prop.IsNull() ? empty_parameters : params_prop.ValueMap(), nullptr);
+              if (!interpreter_context->auth_checker->IsUserAuthorized(owner, prepare_result.privileges)) {
+                throw StreamsException{
+                    "Couldn't execute query '{}' for stream '{}' becuase the owner is not authorized to execute the "
+                    "query!",
+                    query, stream_name};
+              }
+              interpreter->PullAll(&stream);
+            }
+
+            spdlog::trace("Commit transaction in stream '{}'", stream_name);
+            interpreter->CommitTransaction();
+            result.rows.clear();
+          } catch (const std::exception &e) {
+            const std::string_view error(e.what());
+            const bool has_retry_limit_reached = (i + 1) == total_retries;
+            if (error != query::kSerializationErrorMessage || has_retry_limit_reached) {
+              throw;
+            }
```

Comment:
why not define a new type of query that inherits from `QueryRuntimeException` and catch just that one?

Reply:
Yep that sounds reasonable